blog: add walkthrough for Go http-keyvalue-crud example (#815)
wasmCloud/wasmCloud.com: blog/2025-01-23-walkthrough-a-wasmcloud-crud-application-in-go/index.mdx

The only pieces left in our component code are a brief function for JSON validation handling and our empty `main` function.

```go
// JSON validation handling.
func errResponseJSON(w http.ResponseWriter, code int, message string) {
	msg, _ := json.Marshal(CheckResponse{Valid: false, Message: message})
	http.Error(w, string(msg), code)
	w.Header().Set("Content-Type", "application/json")
}

// Since we don't run this program like a CLI, the `main` function is empty. Instead,
// we call handler functions when an HTTP request is received.
func main() {}
```

We've finished walking through the code, and we've already built the `.wasm` binary with `wash build`&mdash;now it's time to look at the application manifest.

## Prepare for deployment

The declarative application manifest in `wadm.yaml` defines the desired state for our application when it is running in a wasmCloud environment. Manifests use the Open Application Model (OAM) standard and will look familiar if you've used Kubernetes. The manifest included with the `http-keyvalue-crud` example looks like this:

> ```yaml

Reviewer: My first thought when I see this and the note at the end about `wash dev` is that they should sort-of be reversed here 🤔 I think that our superpower here around components and autogenerating this manifest is super cool, and the process to do the manual deploy is more relevant _after_ the first deployment if that makes sense.

What do you think? I think I can see the value in outlining the steps here still for the `build + up` style of deployment, but I think it would be better as a secondary note after the initial deployment. Thoughts?

Author: Yeah, on reflection I think you're exactly right. Updated to restructure and run through `wash dev` (and the functionality of the app) first!

Let's walk through the manifest.

* The `metadata` fields provide a name, description, and other optional metadata for the application.
* The `components` fields under `spec` describe the different pieces of our application&mdash;not *just* WebAssembly components, but also capability providers. This application consists of three pieces:
    - The `keyvalue-nats` capability provider, which mediates between the WebAssembly component we built from `main.go` and the actual key-value store we're using&mdash;in this case, the storage built into the [NATS connective layer](/docs/ecosystem/nats/) that is already part of your wasmCloud environment. This provider is fetched as an [OCI artifact](/docs/concepts/packaging). 
    - The `http-server` capability provider, which handles HTTP service, also as an OCI artifact.
    - The `crud` component that we just built, served up straight from the local binary.
* Under the components, we define configuration and links that connect the entities. Links are defined under the `source` (or importer) on the interface in question. (Learn more about linking on the [Linking at runtime](/docs/concepts/linking-components/linking-at-runtime/) overview.)

> When we wrote the code for our WebAssembly component, we didn't worry about how key-value and HTTP services would be rendered. We treated them as abstractions, and now at deployment, we've defined the specific providers that will fulfill those abstract requirements according to the contracts defined in our interfaces. 

Reviewer: ```suggestion
When we wrote the code for our WebAssembly component, we didn't worry about how key-value and HTTP services would be rendered. We treated them as abstractions, and now at deployment, we've defined the specific providers that will fulfill those abstractions with concrete implementations. 
```

Thoughts on this? Trying to cut down on the special words and emphasize abstraction -> concrete.
